File: android-core/src/androidTest/java/com/mparticle/internal/database/services/MessageServiceTest.java
package com.mparticle.internal.database.services;

import android.location.Location;

import com.mparticle.internal.Constants;
import com.mparticle.internal.InternalSession;
import com.mparticle.internal.MessageManager;
import com.mparticle.internal.messages.BaseMPMessage;

import org.json.JSONException;
import org.junit.Before;
import org.junit.Test;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

public class MessageServiceTest extends BaseMPServiceTest {
    Long mpid1, mpid2, mpid3;

    @Before
    public void before() throws Exception {
        mpid1 = ran.nextLong();
        mpid2 = ran.nextLong();
        mpid3 = ran.nextLong();
    }

    @Test
    public void testMessagesForUploadByMpid() throws Exception {
        for (int i = 0; i < 20; i++) {
            MessageService.insertMessage(database, "apiKey", getMpMessage(), mpid1, null, null);
        }
        assertEquals(MessageService.getMessagesForUpload(database, true, mpid1).size(), 20);
        assertEquals(MessageService.getMessagesForUpload(database, true, Constants.TEMPORARY_MPID).size(), 0);
        assertEquals(MessageService.getMessagesForUpload(database).size(), 20);
        for (int i = 0; i < 30; i++) {
            MessageService.insertMessage(database, "apiKey", getMpMessage(), Constants.TEMPORARY_MPID, null, null);
        }
        assertEquals(MessageService.getMessagesForUpload(database, true, mpid1).size(), 20);
        assertEquals(MessageService.getMessagesForUpload(database, true, mpid2).size(), 0);
        assertEquals(MessageService.getMessagesForUpload(database, true, Constants.TEMPORARY_MPID).size(), 30);
        assertEquals(MessageService.getMessagesForUpload(database).size(), 20);
        for (int i = 0; i < 35; i++) {
            MessageService.insertMessage(database, "apiKey", getMpMessage(), mpid2, null, null);
        }
        assertEquals(MessageService.getMessagesForUpload(database, true, mpid1).size(), 20);
        assertEquals(MessageService.getMessagesForUpload(database, true, mpid2).size(), 35);
        assertEquals(MessageService.getMessagesForUpload(database, true, Constants.TEMPORARY_MPID).size(), 30);
        assertEquals(MessageService.getMessagesForUpload(database).size(), 55);

        assertEquals(MessageService.markMessagesAsUploaded(database, Integer.MAX_VALUE), 55);
        assertEquals(MessageService.getMessagesForUpload(database, true, mpid1).size(), 0);
        assertEquals(MessageService.getMessagesForUpload(database, true, mpid2).size(), 0);
        assertEquals(MessageService.getMessagesForUpload(database, true, Constants.TEMPORARY_MPID).size(), 30);
        assertEquals(MessageService.getMessagesForUpload(database).size(), 0);
    }

    @Test
    public void testSessionHistoryByMpid() throws Exception {
        String currentSession = UUID.randomUUID().toString();
        String previousSession = UUID.randomUUID().toString();
        for (int i = 0; i < 20; i++) {
            MessageService.insertMessage(database, "apiKey", getMpMessage(currentSession), mpid1, null, null);
        }
        for (int i = 0; i < 30; i++) {
            MessageService.insertMessage(database, "apiKey", getMpMessage(currentSession), Constants.TEMPORARY_MPID, null, null);
        }
        for (int i = 0; i < 35; i++) {
            MessageService.insertMessage(database, "apiKey", getMpMessage(currentSession), mpid2, null, null);
        }
        assertEquals(MessageService.markMessagesAsUploaded(database, Integer.MAX_VALUE), 55);
        assertEquals(MessageService.getSessionHistory(database, previousSession).size(), 55);
        assertEquals(MessageService.getSessionHistory(database, previousSession, true, mpid1).size(), 20);
        assertEquals(MessageService.getSessionHistory(database, previousSession, true, mpid2).size(), 35);
        assertEquals(MessageService.getSessionHistory(database, previousSession, false, mpid1).size(), 35);
        assertEquals(MessageService.getSessionHistory(database, previousSession, false, Constants.TEMPORARY_MPID).size(), 55);
    }

    @Test
    public void testSessionHistoryAccuracy() throws Exception {
        String currentSession = UUID.randomUUID().toString();
        String previousSession = UUID.randomUUID().toString();
        BaseMPMessage testMessage;
        Long[] mpids = new Long[]{mpid1, mpid2, mpid3};
        Long testMpid;
        Map<String, BaseMPMessage> testMessages = new HashMap<String, BaseMPMessage>();
        for (int i = 0; i < 100; i++) {
            testMpid = mpids[mRandomUtils.randomInt(0, 3)];
            testMessage = getMpMessage(currentSession, testMpid);
            testMessages.put(testMessage.toString(), testMessage);
            MessageService.insertMessage(database, "apiKey",testMessage , testMpid, null, null);
        }
        assertEquals(MessageService.markMessagesAsUploaded(database, Integer.MAX_VALUE), 100);
        List<MessageService.ReadyMessage> readyMessages = MessageService.getSessionHistory(database, previousSession, false, Constants.TEMPORARY_MPID);
        assertEquals(readyMessages.size(), testMessages.size());
        for (MessageService.ReadyMessage readyMessage: readyMessages) {
            BaseMPMessage message = testMessages.get(readyMessage.getMessage());
            assertNotNull(message);
            assertEquals(readyMessage.getMpid(), message.getMpId());
            assertEquals(readyMessage.getMessage(), message.toString());
            assertEquals(readyMessage.getSessionId(), currentSession);
        }

    }

    @Test
    public void testMessageFlow() throws JSONException {
        for (int i = 0; i < 10; i++) {
            MessageService.insertMessage(database, "apiKey", getMpMessage(), 1, "dataplan1", 1);
        }
        List<MessageService.ReadyMessage> messageList = MessageService.getMessagesForUpload(database);
        assertEquals(messageList.size(), 10);
        assertEquals(MessageService.getSessionHistory(database, "123").size(), 0);

        int max = getMaxId(messageList);
        int numUpldated = MessageService.markMessagesAsUploaded(database, max);
        assertEquals(numUpldated, 10);
        assertEquals(MessageService.getMessagesForUpload(database).size(), 0);
        assertEquals(MessageService.getSessionHistory(database, "").size(), 10);
    }

    @Test
    public void testMessageFlowMax() throws JSONException {
        for (int i = 0; i < 110; i++) {
            MessageService.insertMessage(database, "apiKey", getMpMessage(), 1, null, null);
        }
        List<MessageService.ReadyMessage> messages = MessageService.getMessagesForUpload(database);
        assertEquals(messages.size(), 100);
        assertEquals(MessageService.getSessionHistory(database, "").size(), 0);

        int max = getMaxId(messages);
        int numUpdated = MessageService.markMessagesAsUploaded(database, max);
        assertEquals(numUpdated, 100);
        assertEquals(MessageService.getSessionHistory(database, "").size(), 100);

        messages = MessageService.getMessagesForUpload(database);
        max = getMaxId(messages);
        numUpdated = MessageService.markMessagesAsUploaded(database, max);
        assertEquals(numUpdated, 110);
        assertEquals(MessageService.getSessionHistory(database, "").size(), 100);
    }

    @Test
    public void testDeleteOldMessages() throws JSONException {
        String currentSession = UUID.randomUUID().toString();
        String newSession = UUID.randomUUID().toString();
        for (int i = 0; i < 10; i++) {
            MessageService.insertMessage(database, "apiKey", getMpMessage(currentSession), 1, null, null);
        }
        assertEquals(MessageService.markMessagesAsUploaded(database, 10), 10);
        assertEquals(MessageService.getMessagesForUpload(database).size(), 0);

        MessageService.deleteOldMessages(database, currentSession);
        assertEquals(MessageService.getSessionHistory(database, newSession).size(), 10);

        MessageService.deleteOldMessages(database, newSession);
        assertEquals(MessageService.getSessionHistory(database, newSession).size(), 0);
    }

    @Test
    public void testMessagesMaxSize() throws JSONException {
        for (int i = 0; i < 10; i++) {
            MessageService.insertMessage(database, "apiKey", getMpMessage(), 1, "a", 1);
        }
        assertEquals(MessageService.getMessagesForUpload(database).size(), 10);

        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < Constants.LIMIT_MAX_MESSAGE_SIZE; i++) {
            builder.append("ab");
        }
        BaseMPMessage message = new BaseMPMessage.Builder(builder.toString()).build(new InternalSession(), new Location("New York City"), 1);
        MessageService.insertMessage(database, "apiKey", message, 1, "b", 2);

        assertEquals(MessageService.getMessagesForUpload(database).size(), 10);
        for (int i = 0; i < 10; i++) {
            MessageService.insertMessage(database, "apiKey", getMpMessage(), 1, "c", 3);
        }
        assertEquals(MessageService.getMessagesForUpload(database).size(), 20);
    }

    private int getMaxId(List<MessageService.ReadyMessage> messages) {
        int max = 0;
        for (MessageService.ReadyMessage message: messages) {
            if (message.getMessageId() > max) {
                max = message.getMessageId();
            }
        }
        return max;
    }
}

File: android-core/src/androidTest/kotlin/com.mparticle/internal/database/services/MessageServiceTest.kt
package com.mparticle.internal.database.services

import android.location.Location
import com.mparticle.internal.Constants
import com.mparticle.internal.InternalSession
import com.mparticle.internal.database.services.MessageService.ReadyMessage
import com.mparticle.internal.messages.BaseMPMessage
import org.json.JSONException
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import java.util.UUID
import kotlin.properties.Delegates

class MessageServiceTest : BaseMPServiceTest() {
    var mpid1 by Delegates.notNull<Long>()
    var mpid2 by Delegates.notNull<Long>()
    var mpid3 by Delegates.notNull<Long>()

    @Before
    @Throws(Exception::class)
    fun before() {
        mpid1 = ran.nextLong()
        mpid2 = ran.nextLong()
        mpid3 = ran.nextLong()
    }

    @Test
    @Throws(Exception::class)
    fun testMessagesForUploadByMpid() {
        for (i in 0..19) {
            MessageService.insertMessage(
                database,
                "apiKey",
                mpMessage,
                mpid1,
                null,
                null
            )
        }
        Assert.assertEquals(
            MessageService.getMessagesForUpload(
                database,
                true,
                mpid1
            ).size.toLong(),
            20
        )
        Assert.assertEquals(
            MessageService.getMessagesForUpload(
                database,
                true,
                Constants.TEMPORARY_MPID
            ).size.toLong(),
            0
        )
        Assert.assertEquals(MessageService.getMessagesForUpload(database).size.toLong(), 20)
        for (i in 0..29) {
            MessageService.insertMessage(
                database,
                "apiKey",
                mpMessage,
                Constants.TEMPORARY_MPID,
                null,
                null
            )
        }
        Assert.assertEquals(
            MessageService.getMessagesForUpload(
                database,
                true,
                mpid1
            ).size.toLong(),
            20
        )
        Assert.assertEquals(
            MessageService.getMessagesForUpload(
                database,
                true,
                mpid2
            ).size.toLong(),
            0
        )
        Assert.assertEquals(
            MessageService.getMessagesForUpload(
                database,
                true,
                Constants.TEMPORARY_MPID
            ).size.toLong(),
            30
        )
        Assert.assertEquals(MessageService.getMessagesForUpload(database).size.toLong(), 20)
        for (i in 0..34) {
            MessageService.insertMessage(
                database,
                "apiKey",
                mpMessage,
                mpid2,
                null,
                null
            )
        }
        Assert.assertEquals(
            MessageService.getMessagesForUpload(
                database,
                true,
                mpid1
            ).size.toLong(),
            20
        )
        Assert.assertEquals(
            MessageService.getMessagesForUpload(
                database,
                true,
                mpid2
            ).size.toLong(),
            35
        )
        Assert.assertEquals(
            MessageService.getMessagesForUpload(
                database,
                true,
                Constants.TEMPORARY_MPID
            ).size.toLong(),
            30
        )
        Assert.assertEquals(MessageService.getMessagesForUpload(database).size.toLong(), 55)
        Assert.assertEquals(
            MessageService.markMessagesAsUploaded(database, Int.MAX_VALUE).toLong(),
            55
        )
        Assert.assertEquals(
            MessageService.getMessagesForUpload(
                database,
                true,
                mpid1
            ).size.toLong(),
            0
        )
        Assert.assertEquals(
            MessageService.getMessagesForUpload(
                database,
                true,
                mpid2
            ).size.toLong(),
            0
        )
        Assert.assertEquals(
            MessageService.getMessagesForUpload(
                database,
                true,
                Constants.TEMPORARY_MPID
            ).size.toLong(),
            30
        )
        Assert.assertEquals(MessageService.getMessagesForUpload(database).size.toLong(), 0)
    }

    @Test
    @Throws(Exception::class)
    fun testSessionHistoryByMpid() {
        val currentSession = UUID.randomUUID().toString()
        val previousSession = UUID.randomUUID().toString()
        for (i in 0..19) {
            MessageService.insertMessage(
                database,
                "apiKey",
                getMpMessage(currentSession),
                mpid1,
                null,
                null
            )
        }
        for (i in 0..29) {
            MessageService.insertMessage(
                database,
                "apiKey",
                getMpMessage(currentSession),
                Constants.TEMPORARY_MPID,
                null,
                null
            )
        }
        for (i in 0..34) {
            MessageService.insertMessage(
                database,
                "apiKey",
                getMpMessage(currentSession),
                mpid2,
                null,
                null
            )
        }
        Assert.assertEquals(
            MessageService.markMessagesAsUploaded(database, Int.MAX_VALUE).toLong(),
            55
        )
        Assert.assertEquals(
            MessageService.getSessionHistory(
                database,
                previousSession
            ).size.toLong(),
            55
        )
        Assert.assertEquals(
            MessageService.getSessionHistory(
                database,
                previousSession,
                true,
                mpid1
            ).size.toLong(),
            20
        )
        Assert.assertEquals(
            MessageService.getSessionHistory(
                database,
                previousSession,
                true,
                mpid2
            ).size.toLong(),
            35
        )
        Assert.assertEquals(
            MessageService.getSessionHistory(
                database,
                previousSession,
                false,
                mpid1
            ).size.toLong(),
            35
        )
        Assert.assertEquals(
            MessageService.getSessionHistory(
                database,
                previousSession,
                false,
                Constants.TEMPORARY_MPID
            ).size.toLong(),
            55
        )
    }

    @Test
    @Throws(Exception::class)
    fun testSessionHistoryAccuracy() {
        val currentSession = UUID.randomUUID().toString()
        val previousSession = UUID.randomUUID().toString()
        var testMessage: BaseMPMessage
        val mpids = arrayOf(mpid1, mpid2, mpid3)
        var testMpid: Long?
        val testMessages: MutableMap<String, BaseMPMessage> = HashMap()
        for (i in 0..99) {
            testMpid = mpids[mRandomUtils.randomInt(0, 3)]
            testMessage = getMpMessage(currentSession, testMpid)
            testMessages[testMessage.toString()] = testMessage
            MessageService.insertMessage(
                database,
                "apiKey",
                testMessage,
                testMpid,
                null,
                null
            )
        }
        Assert.assertEquals(
            MessageService.markMessagesAsUploaded(database, Int.MAX_VALUE).toLong(),
            100
        )
        val readyMessages = MessageService.getSessionHistory(
            database,
            previousSession,
            false,
            Constants.TEMPORARY_MPID
        )
        Assert.assertEquals(readyMessages.size.toLong(), testMessages.size.toLong())
        for (readyMessage in readyMessages) {
            val message = testMessages[readyMessage.message]
            Assert.assertNotNull(message)
            Assert.assertEquals(readyMessage.mpid, message?.mpId)
            Assert.assertEquals(readyMessage.message, message.toString())
            Assert.assertEquals(readyMessage.sessionId, currentSession)
        }
    }

    @Test
    @Throws(JSONException::class)
    fun testMessageFlow() {
        for (i in 0..9) {
            MessageService.insertMessage(database, "apiKey", mpMessage, 1, "dataplan1", 1)
        }
        val messageList = MessageService.getMessagesForUpload(database)
        Assert.assertEquals(messageList.size.toLong(), 10)
        Assert.assertEquals(MessageService.getSessionHistory(database, "123").size.toLong(), 0)
        val max = getMaxId(messageList)
        val numUpldated = MessageService.markMessagesAsUploaded(database, max)
        Assert.assertEquals(numUpldated.toLong(), 10)
        Assert.assertEquals(MessageService.getMessagesForUpload(database).size.toLong(), 0)
        Assert.assertEquals(MessageService.getSessionHistory(database, "").size.toLong(), 10)
    }

    @Test
    @Throws(JSONException::class)
    fun testMessageFlowMax() {
        for (i in 0..109) {
            MessageService.insertMessage(database, "apiKey", mpMessage, 1, null, null)
        }
        var messages = MessageService.getMessagesForUpload(database)
        Assert.assertEquals(messages.size.toLong(), 100)
        Assert.assertEquals(MessageService.getSessionHistory(database, "").size.toLong(), 0)
        var max = getMaxId(messages)
        var numUpdated = MessageService.markMessagesAsUploaded(database, max)
        Assert.assertEquals(numUpdated.toLong(), 100)
        Assert.assertEquals(MessageService.getSessionHistory(database, "").size.toLong(), 100)
        messages = MessageService.getMessagesForUpload(database)
        max = getMaxId(messages)
        numUpdated = MessageService.markMessagesAsUploaded(database, max)
        Assert.assertEquals(numUpdated.toLong(), 110)
        Assert.assertEquals(MessageService.getSessionHistory(database, "").size.toLong(), 100)
    }

    @Test
    @Throws(JSONException::class)
    fun testDeleteOldMessages() {
        val currentSession = UUID.randomUUID().toString()
        val newSession = UUID.randomUUID().toString()
        for (i in 0..9) {
            MessageService.insertMessage(
                database,
                "apiKey",
                getMpMessage(currentSession),
                1,
                null,
                null
            )
        }
        Assert.assertEquals(MessageService.markMessagesAsUploaded(database, 10).toLong(), 10)
        Assert.assertEquals(MessageService.getMessagesForUpload(database).size.toLong(), 0)
        MessageService.deleteOldMessages(database, currentSession)
        Assert.assertEquals(
            MessageService.getSessionHistory(database, newSession).size.toLong(),
            10
        )
        MessageService.deleteOldMessages(database, newSession)
        Assert.assertEquals(MessageService.getSessionHistory(database, newSession).size.toLong(), 0)
    }

    @Test
    @Throws(JSONException::class)
    fun testMessagesMaxSize() {
        for (i in 0..9) {
            MessageService.insertMessage(database, "apiKey", mpMessage, 1, "a", 1)
        }
        Assert.assertEquals(MessageService.getMessagesForUpload(database).size.toLong(), 10)
        val builder = StringBuilder()
        for (i in 0 until Constants.LIMIT_MAX_MESSAGE_SIZE) {
            builder.append("ab")
        }
        val message = BaseMPMessage.Builder(builder.toString())
            .build(InternalSession(), Location("New York City"), 1)
        MessageService.insertMessage(database, "apiKey", message, 1, "b", 2)
        Assert.assertEquals(MessageService.getMessagesForUpload(database).size.toLong(), 10)
        for (i in 0..9) {
            MessageService.insertMessage(database, "apiKey", mpMessage, 1, "c", 3)
        }
        Assert.assertEquals(MessageService.getMessagesForUpload(database).size.toLong(), 20)
    }

    private fun getMaxId(messages: List<ReadyMessage>): Int {
        var max = 0
        for (message in messages) {
            if (message.messageId > max) {
                max = message.messageId
            }
        }
        return max
    }
}
